### DISCUSSION TITLE
ObjectMapper is better to have as a field rather than a local variable

### DISCUSSION THREAD

**Discussion by dgroup** [dgroup/arch4u-pmd#30] [Ideas]
### Summary
| Key            |  Value |
|:-----------|:------|
|  Rule           | AvoidUsingObjectMapperAsALocalVariable   |
|Ruleset        | [io/github/dgroup/arch4u/pmd/arch4u-ruleset.xml](https://github.com/dgroup/arch4u-pmd/blob/master/src/main/resources/io/github/dgroup/arch4u/pmd/arch4u-ruleset.xml)|
|Category     | Performance |
|Framework | [com.fasterxml.jackson.core:jackson-databind](https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind) |
|Since           |    0.1.0    |
|Incidents in past at my experience | 1 |

### Why?
ObjectMapper is better to have as a field rather than a local variable due to [performance reasons](https://github.com/dgroup/arch4u-pmd/issues/99).
It is allowed to be declared in fields, constructors, and initialization blocks.

```java
class Foo {
    ObjectMapper objectMapper = new ObjectMapper(); //ok

    {
        objectMapper = new ObjectMapper();  //ok
    }

    Foo (ObjectMapper om) {                 //ok
        objectMapper = new ObjectMapper();  //ok
    }

    void bar(ObjectMapper om) {                                  //violation
        ObjectMapper mapper = om;                                //violation
        mapper = new ObjectMapper();                             //violation
        String str = new ObjectMapper().writeValueAsString(obj); //violation
    }
}
```

**Comment by davidrupp** [comment]
Hello. Can you share some details about the "incident in past" you've experienced that led to this conversation?

**Comment by Cimballi** [comment]
What about adding a similar rule but for Gson ?
